numba/targets/arrayobj.py

```diff
@@ -3756,70 +3756,95 @@ def take_impl(a, indices):
+    step = _step if _step is not None else 1
+    if _stop is None:
+        start, stop = 0, _start
+    else:
+        start, stop = _start, _stop
 
-@lower_builtin(np.arange, types.Number, types.Number,
-               types.Number)
-def numpy_arange_3(context, builder, sig, args):
-    dtype = as_dtype(sig.return_type.dtype)
-
-    def arange(start, stop, step):
-        return np.arange(start, stop, step, dtype)
+    nitems_c = (stop - start) / step
+    nitems_r = int(math.ceil(nitems_c.real))
+    nitems_i = int(math.ceil(nitems_c.imag))
+    nitems = max(min(nitems_i, nitems_r), 0)
+    arr = np.empty(nitems, dtype)
+    val = start
+    for i in range(nitems):
+        arr[i] = val
+        val += step
+    return arr
 
-    res = context.compile_internal(builder, arange, sig, args)
-    return impl_ret_new_ref(context, builder, sig.return_type, res)
 
+@register_jitable
+def _arange_impl_real(_start, _stop, _step, dtype):
+    step = _step if _step is not None else 1
+    if _stop is None:
+        start, stop = 0, _start
+    else:
+        start, stop = _start, _stop
+
+    nitems_r = int(math.ceil((stop - start) / step))
+    nitems = max(nitems_r, 0)
+    arr = np.empty(nitems, dtype)
+    val = start
+    for i in range(nitems):
+        arr[i] = val
+        val += step
+    return arr
+
+
+@overload(np.arange)
+def np_arange(start, stop=None, step=None, dtype=None):
+    if isinstance(stop, types.Optional):
+        stop = stop.type
+    if isinstance(step, types.Optional):
+        step = step.type
+    if isinstance(dtype, types.Optional):
+        dtype = dtype.type
+
+    if stop is None:
+        stop = types.none
+    if step is None:
+        step = types.none
+    if dtype is None:
+        dtype = types.none
+
+    if (not isinstance(start, types.Number) or
+        not isinstance(stop, (types.NoneType, types.Number)) or
+        not isinstance(step, (types.NoneType, types.Number)) or
+            not isinstance(dtype, (types.NoneType, types.DTypeSpec))):
+
+        return
+
+    if isinstance(dtype, types.NoneType):
+        true_dtype = _arange_dtype(start, stop, step)
+    else:
+        true_dtype = dtype.dtype
 
-@lower_builtin(np.arange, types.Number, types.Number,
-               types.Number, types.DTypeSpec)
-def numpy_arange_4(context, builder, sig, args):
-
-    if any(isinstance(a, types.Complex) for a in sig.args):
-        def arange(start, stop, step, dtype):
-            nitems_c = (stop - start) / step
-            nitems_r = math.ceil(nitems_c.real)
-            nitems_i = math.ceil(nitems_c.imag)
-            nitems = max(min(nitems_i, nitems_r), 0)
-            arr = np.empty(nitems, dtype)
-            val = start
-            for i in range(nitems):
-                arr[i] = val
-                val += step
-            return arr
+    use_complex = any([isinstance(x, types.Complex)
+                       for x in (start, stop, step)])
+    if use_complex:
```

Comment:
This would benefit from leaning on branch pruning (removing dead branches based on a condition evaluated at compile time). Currently the condition has to be a binop, but that's a small price to pay to cut down the code. Perhaps take a look at: https://github.com/stuartarchibald/numba/commit/9405791041bd6c8956be5b4d53a5f60bee63767b, the `use_complex` freevar is a compile time constant and depending on its value the True/False branch is "pruned" from the code, this permitting specialisation on type without code duplication.

Reply:
Merged your commit.

Reply:
great, glad it helped.